src/lang/std/engineConnection.ts

```diff
@@ -1179,23 +1331,36 @@ export class EngineCommandManager {
               this.handleFailedModelingCommand(message.request_id, message)
             }
           }
-        })
-      },
-      onNewTrack: ({ mediaStream }) => {
-        console.log('received track', mediaStream)
+        }) as EventListener)
 
-        mediaStream.getVideoTracks()[0].addEventListener('mute', () => {
-          console.log('peer is not sending video to us')
-          // this.engineConnection?.close()
-          // this.engineConnection?.connect()
-        })
+        this.engineConnection?.addEventListener(
+          EngineConnectionEvents.NewTrack,
+          (({ detail: { mediaStream } }: CustomEvent<NewTrackArgs>) => {
+            console.log('received track', mediaStream)
 
-        setMediaStream(mediaStream)
-      },
-    })
+            mediaStream.getVideoTracks()[0].addEventListener('mute', () => {
+              if (this.engineConnection) {
+                this.engineConnection.state = {
+                  type: EngineConnectionStateType.Disconnecting,
+                  value: {
+                    type: DisconnectingType.Error,
+                    value: {
+                      error: ConnectionError.LostVideoStream,
+                    },
+                  },
+                }
+              }
+            })
```

Comment:
Muting the stream causes a lost video stream error?

Reply:
This logic was already there before; it's an early signal that the stream is lost (the video goes mute). I just added the event 'sall. It's actually pretty effective way to see an early stream close.